extensions/amp-bind/0.1/bind-impl.js

```diff
@@ -0,0 +1,243 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {evaluateBindExpr} from './bind-expr';
+import {getMode} from '../../../src/mode';
+import {user} from '../../../src/log';
+import {vsyncFor} from '../../../src/vsync';
+
+const TAG_ = 'AMP-BIND';
+
+/**
+ * @typedef {{
+ *   property: !string,
+ *   expression: !string,
+ *   element: !Element
+ * }}
+ */
+let BindingDef;
+
+export class Bind {
+  /**
+   * @param {!../../../src/service/ampdoc-impl.AmpDoc} ampdoc
+   */
+  constructor(ampdoc) {
+    /** @const {!../../../src/service/ampdoc-impl.AmpDoc} */
+    this.ampdoc = ampdoc;
+
+    /** @const {!Array<BindingDef>} */
+    this.bindings_ = [];
+
+    /** @const {!Object} */
+    this.scope_ = Object.create(null);
+
+    /** @const {!../../../src/service/vsync-impl.Vsync} */
+    this.vsync_ = vsyncFor(ampdoc.win);
+
+    /** @const {!Array<string>} */
+    this.protocolWhitelist_ = ['http', 'https'];
+
+    this.ampdoc.whenBodyAvailable().then(body => {
+      this.bindings_ = this.scanForBindings_(body);
+
+      // Trigger verify-only digest in development.
+      if (getMode().development) {
+        this.digest_(true);
+      }
+    });
+  }
+
+  /**
+   * @param {!Object} state
+   * @param {boolean=} opt_skipDigest
+   */
+  setState(state, opt_skipDigest) {
+    Object.assign(this.scope_, state);
+
+    if (!opt_skipDigest) {
+      this.digest_();
+    }
+  }
+
+  /**
+   * Scans children for attributes that conform to bind syntax and returns
+   * all bindings.
+   * @param {!Element} body
+   * @return {!Array<BindingDef>}
+   * @private
+   */
+  scanForBindings_(body) {
+    const bindings = [];
+    const elements = body.getElementsByTagName('*');
```

Comment:
Can we force all elements with bindings to have a `bind` attribute? Or some other way to distinguish them without checking every element?